### PR TITLE
[CDAP-7252] Feature/ui namespace dropdown

### PR DESCRIPTION
**Pull request by PGuimarais**
- Now uses React-Router v4.0 for routing
- CDAP component now polls for namespaces before mounting
- Adds Redux Store for list of namespaces and currently selected namespace
- Implements namespace dropdown

screen shot 2016-09-19 at 12 05 16 pm

Bamboo Build: 
http://builds.cask.co/browse/CDAP-DRC4487

### CODE REVIEW

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 1]
```diff
@@ -19,21 +19,55 @@ import { Dropdown, DropdownMenu, DropdownItem } from 'reactstrap';
...
+      namespaceOpen : false,
+      currentNamespace: Store.getState().selectedNamespace
     };
     this.logout = this.logout.bind(this);
     this.toggleSettingsDropdown = this.toggleSettingsDropdown.bind(this);
+    this.namespaceList = [];
+    this.namespaceMap = '';
+    this.toggleNamespaceDropdown = this.toggleNamespaceDropdown.bind(this);
+    this.loadNamespacesFromStore = this.loadNamespacesFromStore.bind(this);
+    this.selectNamespaceAndCloseDropdown = this.selectNamespaceAndCloseDropdown.bind(this);
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: why are these needed?

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 2]
```diff
@@ -52,6 +86,61 @@ export default class HeaderActions extends Component {
...
+
+      //If there is no current namespace, set it to default if possible
+      if(!this.state.currentNamespace && item.name === 'default'){
+        this.defaultNamespace = item.name;
+      }
+
+      var checkClass = classNames({ "fa fa-check selected-namespace-check": current === item.name });
+
+      var handler = this.selectNamespaceAndCloseDropdown.bind(this, item.name);
+      var check = <span className={checkClass}></span>;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: don't use `var` unless there's a need

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 3]
```diff
@@ -133,8 +223,23 @@ export default class HeaderActions extends Component {
               </DropdownMenu>
             </Dropdown>
           </div>
-          <div className="navbar-item namespace-dropdown dropdown">
-            <span> Namespace </span>
+          <div className="navbar-item namespace-dropdown">
+            <Dropdown isOpen={this.state.namespaceOpen}
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: isOpen should be on a newline with only single tab

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 4]
```diff
@@ -133,8 +223,23 @@ export default class HeaderActions extends Component {
...
+                      toggle={this.toggleNamespaceDropdown}
+            >
+              <div className="current-namespace"
+                   onClick={this.toggleNamespaceDropdown}
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: same here

File: cdap-ui/app/cdap/components/HeaderNavbarList/index.js [hunk 5]
```diff
@@ -16,15 +16,18 @@
 import React, {PropTypes} from 'react';
 import {Link} from 'react-router';
 
-export default function HeaderNavbarList({list}) {
+export default function HeaderNavbarList({list, pathname}) {
+
+///\navbaritem.linkTo
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderNavbarList/index.js]: remove

File: cdap-ui/app/cdap/components/HeaderNavbarList/index.js [hunk 6]
```diff
@@ -16,15 +16,18 @@
...
+
   return (
     <ul className="navbar-list">
       {
         list.map((navbaritem, index) => {
           return (
             <li key={index}>
               <Link
-                to={navbaritem.linkTo}
+                to={`${pathname}/` + navbaritem.linkTo}
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/HeaderNavbarList/index.js]: if this is a string, you can simply do ``${pathname}/${navbaritem.linkTo}``

File: cdap-ui/app/cdap/components/Home/index.js [hunk 7]
```diff
@@ -14,19 +14,20 @@
...
+import Store from '../../services/store/store.js';
 
 require('./Home.less');
 
-export default class Home extends Component {
+class Home extends Component {
   constructor(props) {
     super(props);
-
+    this.props = props;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Home/index.js]: I don't think this is neeeded. The props should already be initialized within `super(props)`

File: cdap-ui/app/cdap/components/Home/index.js [hunk 8]
```diff
@@ -63,6 +64,27 @@ export default class Home extends Component {
     this.search();
   }
 
+  findNamespace(){
+    var namespaces = Store.getState().namespaces;
+    var name = this.props.params.namespace;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Home/index.js]: don't use `var`

File: cdap-ui/app/cdap/components/Home/index.js [hunk 9]
```diff
@@ -63,6 +64,27 @@ export default class Home extends Component {
...
+
+    for(var i = 0; i < namespaces.length; i++){
+      if(namespaces[i].name === name){
+        return true;
+      }
+    }
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Home/index.js]: should be able to use lodash to simplify this

File: cdap-ui/app/cdap/main.js [hunk 10]
```diff
@@ -36,17 +33,83 @@ import SplashScreen from './components/SplashScreen';
...
+
+          Store.dispatch({
+            type: 'UPDATE_NAMESPACES',
+            payload: {
+              namespaces : res
+            }
+          });
+
+          Store.dispatch({
+            type: 'SELECT_NAMESPACE',
```
Comment (elfenheart) [inline cdap-ui/app/cdap/main.js]: this will get triggered when the list of namespace gets updated. Is that intentional?

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 11]
```diff
@@ -17,23 +17,59 @@
...
 
 export default class HeaderActions extends Component {
 
   constructor(props) {
     super(props);
     this.state = {
       settingsOpen : false,
-      name : cookie.load('CDAP_Auth_User')
+      name : cookie.load('CDAP_Auth_User'),
+      namespaceList : [],
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js:35]: This should come from the store. The default state should be from the store.

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 12]
```diff
@@ -17,23 +17,59 @@
...
+    }
+
+    //Load updated data into the namespace dropdown
+    Store.subscribe(() => {
+
+      this.namespaceList = Store.getState().namespaces;
+      var namespaceToUpdate = Store.getState().selectedNamespace;
+
+      if(this.namespaceList){
+        this.loadNamespacesFromStore(this.namespaceList, () => {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: Why do we need to have the second argument as an empty callback function? Any reason why we can't have `namespaceToUpdate` as the second argument here?

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 13]
```diff
@@ -52,6 +88,58 @@ export default class HeaderActions extends Component {
...
+  }
+
+  loadNamespacesFromStore (res, current) {
+
+    if(typeof res === 'undefined'){
+      return;
+    }
+
+    if(current){
+      current = current();
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: If we pass the current as string then we shouldn't probably do this. In general it would be nice if we don't change the type of a function argument that is coming. Originally this is coming in as function but here we are modifying it as string or something else. It would be better if we avoid this

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 14]
```diff
@@ -52,6 +88,58 @@ export default class HeaderActions extends Component {
     });
   }
 
+  toggleNamespaceDropdown(){
+    this.setState({
+      namespaceOpen : !this.state.namespaceOpen
+    });
+  }
+
+  loadNamespacesFromStore (res, current) {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: Probably better names may be? `namespaceList` , `currentNamespace`?

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 15]
```diff
@@ -52,6 +88,58 @@ export default class HeaderActions extends Component {
...
+          <Link to={`/ns/${item.name}`}>
+            {check}
+            {item.name}
+          </Link>
+        </div>
+      );
+    });
+  }
+
+  selectNamespaceAndCloseDropdown(name){
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: I think `selectNamespace` here would be sufficient . Closing the dropdown is something that I think is implicit. Doesn't need to be in the function name

File: cdap-ui/app/cdap/components/HeaderActions/index.js [hunk 16]
```diff
@@ -52,6 +88,58 @@ export default class HeaderActions extends Component {
...
+    this.setState({
+      currentNamespace : name,
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderActions/index.js]: This should be set in the store's subscribe function right? Any specific reason why we are doing it here?

File: cdap-ui/app/cdap/components/HeaderNavbarList/index.js [hunk 17]
```diff
@@ -13,28 +13,58 @@
...
-              >
-                {navbaritem.title}
-              </Link>
-            </li>
-          );
-        })
-      }
-    </ul>
-  );
+export default class HeaderNavbarList extends Component {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/HeaderNavbarList/index.js]: We don't need to use es6 class here. We should use `react-redux`'s `connect` function. It automatically takes care of subscribing, binding it to properties, updating the state and unsubscribing from the store on unMounting the component. The boilerplate of doing these are already in there.

File: cdap-ui/app/cdap/components/Home/index.js [hunk 18]
```diff
@@ -61,6 +61,48 @@ export default class Home extends Component {
...
+  componentWillMount(){
+    if(this.findNamespace()){
+      Store.dispatch({
+        type : 'SELECT_NAMESPACE',
+        payload : {
+          selectedNamespace : this.props.params.namespace
+        }
+      });
+    } else {
+      //FIXME: Must redirect if passed invalid namespace
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/Home/index.js:91]: Did this pass eslint in build?

File: cdap-ui/app/cdap/main.js [hunk 19]
```diff
@@ -36,17 +33,91 @@ import SplashScreen from './components/SplashScreen';
...
     super(props);
     this.props = props;
     this.version = '4.0.0';
+    this.namespace = Store.getState().selectedNamespace;
+    this.namespaceList = Store.getState().namespaceList;
+    this.pathname = props.pathname;
+    this.render = this.render.bind(this);
+    Store.subscribe(this.render);
+    this.state = {
+      selectedNamespace : ''
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/main.js]: The default value should come from the store.

File: cdap-ui/app/cdap/services/store/actions.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+export function updateNamespaces(namespaces) {
+  return {
+    type: 'UPDATE_NAMESPACES',
+    payload: {
+      namespaceList : namespaces
+    }
+  };
+}
+
+export function selectNamespace(namespace) {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/services/store/actions.js]: Are these being used anywhere?

**Review by ajainarayanan — CHANGES_REQUESTED** [review]

File: cdap-ui/app/cdap/components/Home/index.js (hunk 18 above)
Comment (PGuimarais) [reply cdap-ui/app/cdap/components/Home/index.js:91]: yes

**Comment by PGuimarais** [comment]
@ajainarayanan @elfenheart addressed comments

**Comment by elfenheart** [comment]
screen shot 2016-09-21 at 2 11 20 pm
1. highlight should cover the entire row. 
2. The click handler should also be on the entire row, not just on the text
3. The active Navbar links should get highlighted

**Comment by elfenheart** [comment]
nit style: add space after `if`s and before the opening curly brace

**Comment by elfenheart** [comment]
👍  LGTM
